framework/src/base/FEProblemBase.C

```diff
@@ -2016,25 +2016,35 @@ void
 FEProblemBase::prepareMaterials(SubdomainID blk_id, THREAD_ID tid)
 {
   std::set<MooseVariable *> needed_moose_vars;
+  std::set<unsigned int> needed_mat_props;
 
   if (_all_materials.hasActiveBlockObjects(blk_id, tid))
+  {
     _all_materials.updateVariableDependency(needed_moose_vars, tid);
+    _all_materials.updateBlockMatPropDependency(blk_id, needed_mat_props, tid);
+  }
 
   const std::set<BoundaryID> & ids = _mesh.getSubdomainBoundaryIds(blk_id);
   for (const auto & id : ids)
+  {
     _materials.updateBoundaryVariableDependency(id, needed_moose_vars, tid);
+    _materials.updateBoundaryMatPropDependency(id, needed_mat_props, tid);
+  }
 
   const std::set<MooseVariable *> & current_active_elemental_moose_variables = getActiveElementalMooseVariables(tid);
   needed_moose_vars.insert(current_active_elemental_moose_variables.begin(), current_active_elemental_moose_variables.end());
 
-  if (!needed_moose_vars.empty())
-    setActiveElementalMooseVariables(needed_moose_vars, tid);
+  const std::set<unsigned int> & current_active_material_properties = getActiveMaterialProperties(tid);
+  needed_mat_props.insert(current_active_material_properties.begin(), current_active_material_properties.end());
+
+  setActiveElementalMooseVariables(needed_moose_vars, tid);
+  setActiveMaterialProperties(needed_mat_props, tid);
 }
 
 void
 FEProblemBase::reinitMaterials(SubdomainID blk_id, THREAD_ID tid, bool swap_stateful)
 {
-  if (_all_materials.hasActiveBlockObjects(blk_id, tid))
+  if (_all_materials.hasActiveBlockObjects(blk_id, tid) && this->hasActiveMaterialProperties(tid))
```

Comment:
I can see the logic in switching the order for when `hasActiveMaterialProperties` evaluates to `false` and we use "and", but I don't see how the "or" would be a good idea. Because it would defeat the whole purpose of the PR if `hasActiveMaterialProperties` evaluated to 'false` and `hasActiveBlockObjects` evaluated to `true` and we used an "or" comparison.